turbo/trie/gen_struct_step.go

```diff
@@ -267,11 +276,21 @@ func GenStructStep(
 				}
 			}
 
+			var doProof bool
+			if wantProof != nil {
+				if maxLen > 0 && wantProof(curr[:maxLen]) {
+					doProof = true
+				}
+				if len(succ) == 0 && maxLen == 0 && cutoff {
+					doProof = true
+				}
+			}
+
 			if trace {
 				e.printTopHashes(curr[:maxLen], 0, groups[maxLen])
 			}
-			if retain(curr[:maxLen]) {
-				if err := e.branch(groups[maxLen]); err != nil {
+			if retain(curr[:maxLen]) || doProof {
+				if err := e.branchEx(groups[maxLen], doProof); err != nil {
```

Comment:
- Is it possible to call `e.TopHashes()` before/after `e.branch()` to get info which you need? (instead of introducing `e.branchEx`)
-

Reply:
I'll look at it. I need the trie nodes before they are hashed and it looked like topHashes() was only giving me the output of that operation.

Reply:
@mmontour1306 , yes, but it has stack of nodes `HashBuilder.nodeStack` just no method like `TopNodes()` to return top of stack. Also maybe need read it before hashing, not after.

Reply:
Anyway, not very important. I just trying to keep GenStruct step simpler. It's already complex. So, if you have any idea about it - you are welcome.

Reply:
In my latest version I have removed the "e.branchEx" stuff, restoring the original function signatures. Instead of passing a "doProof" flag into those functions, I have instead introduced a new function which is called ahead of time to set a flag telling hashbuilder that it should collect the proof data the next time it generates a hash (for a branch node or for the account leaf node). This is the least invasive method I can think of.

The various internal stacks do not appear to ever contain the specific data structures I need for the proof. The raw data is there and it would be possible to construct another copy of the node structure elsewhere, but this would involve duplicating a lot of the hashbuilder code in that other location. The data I need is what's fed into the hashing function inside hashbuilder as it constructs the RLP-encoded data structures.